### PR TITLE
Introduce stricter asserts for mapping into/out of generic contexts

### PR DESCRIPTION
**Pull request by swiftix**

### CODE REVIEW

**Comment by swiftix** [comment]
@swift-ci please smoke test and merge

**Comment by swiftix** [comment]
@swift-ci please clean test Linux

**Comment by swiftix** [comment]
@swift-ci please smoke test

**Comment by swiftix** [comment]
@swift-ci please clean smoke test Linux

File: lib/SIL/SILFunction.cpp [hunk 1]
```diff
@@ -162,11 +162,15 @@ bool SILFunction::shouldOptimize() const {
...
 
 Type SILFunction::mapTypeIntoContext(Type type) const {
+  assert((!type->hasTypeParameter() || getGenericEnvironment()) &&
+         "SILFunction should have a generic environment");
   return GenericEnvironment::mapTypeIntoContext(
       getGenericEnvironment(), type);
 }
 
 SILType SILFunction::mapTypeIntoContext(SILType type) const {
+  assert((!type.hasTypeParameter() || getGenericEnvironment()) &&
```
Comment (slavapestov) [inline lib/SIL/SILFunction.cpp]: This is the only place you need the assert. The static forms of mapTypeIn/OutOfContext already assert that the type is fully concrete if the given environment is null. The issue is there's no static method equivalent for the SILType version

**Review by slavapestov — APPROVED** [review]

File: lib/SIL/SILFunction.cpp (hunk 1 above)
Comment (swiftix) [inline lib/SIL/SILFunction.cpp]: They do not assert that the type is fully concrete if the environment is null, AFAIK:

```c++
Type GenericEnvironment::mapTypeIntoContext(GenericEnvironment *env,
                                            Type type) {
  assert(!type->hasArchetype() && "already have a contextual type");

  if (!env)
    return type.substDependentTypesWithErrorTypes();

  return env->mapTypeIntoContext(type);
}

Type
GenericEnvironment::mapTypeOutOfContext(GenericEnvironment *env,
                                        Type type) {
  assert(!type->hasTypeParameter() && "already have an interface type");

  if (!env)
    return type.substDependentTypesWithErrorTypes();

  return env->mapTypeOutOfContext(type);
}
```

Comment (swiftix) [inline lib/SIL/SILFunction.cpp]: `type.substDependentTypesWithErrorTypes` would probably return an error type, I guess.

Comment (slavapestov) [inline lib/SIL/SILFunction.cpp]: I'm planning on replacing that function with an assert soon.

**Pull request closed without merging**